### PR TITLE
Add unit tests and code generation routines

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**

### CODE REVIEW

File: test/Microsoft.Framework.WebEncoders.Tests/CodePointFilterTests.cs [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+
+            // Act
+            var retVal = filter.GetAllowedCodePoints().OrderBy(i => i).ToArray();
+
+            // Assert
+            Assert.Equal<int>(expected, retVal);
+        }
+
+        // a code point filter which allows only odd code points through
+        private sealed class OddCodePointFilter : ICodePointFilter
```
Comment (Eilon) [inline test/Microsoft.Framework.WebEncoders.Tests/CodePointFilterTests.cs]: This is indeed an *odd* code point filter :trollface:

File: test/Microsoft.Framework.WebEncoders.Tests/CommonTestEncoder.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        /// Writes "UrlEncode[[value]]".
+        /// </summary>
+        public void UrlEncode(char[] value, int startIndex, int charCount, TextWriter output)
+        {
+            EncodeCore(value, startIndex, charCount, output);
+        }
+
+        private static string EncodeCore(string value, [CallerMemberName] string encodeType = null)
+        {
+            return String.Format(CultureInfo.InvariantCulture, "{0}[[{1}]]", encodeType, value);
```
Comment (Eilon) [inline test/Microsoft.Framework.WebEncoders.Tests/CommonTestEncoder.cs]: Style: String -> string, Char -> char (throughout)

File: test/Microsoft.Framework.WebEncoders.Tests/EncoderExtensionsTests.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+using Xunit;
+
+namespace Microsoft.Framework.WebEncoders
+{
+    public class EncoderExtensionsTests
+    {
+        [Fact]
+        public void HtmlEncode_ParameterChecks()
+        {
+            Assert.Throws<ArgumentNullException>(() => EncoderExtensions.HtmlEncode(null, "Hello!", new StringWriter()));
```
Comment (Eilon) [inline test/Microsoft.Framework.WebEncoders.Tests/EncoderExtensionsTests.cs]: I think there's a param to test the param name of the exception, no?

File: test/Microsoft.Framework.WebEncoders.Tests/EncoderExtensionsTests.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        }
+
+        [Fact]
+        public void HtmlEncode_PositiveTestCase()
+        {
+            // Arrange
+            IHtmlEncoder encoder = new HtmlEncoder(UnicodeBlocks.All);
+            StringWriter writer = new StringWriter();
```
Comment (Eilon) [inline test/Microsoft.Framework.WebEncoders.Tests/EncoderExtensionsTests.cs]: `var` all day long.

File: test/Microsoft.Framework.WebEncoders.Tests/HtmlEncoderTests.cs [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+        [Fact]
+        public void HtmlEncode_AllRangesAllowed_StillEncodesForbiddenChars_Extended()
+        {
+            // Arrange
+            HtmlEncoder encoder = new HtmlEncoder(UnicodeBlocks.All);
+
+            // Act & assert - BMP chars
+            for (int i = 0; i <= 0xFFFF; i++)
+            {
```
Comment (Eilon) [inline test/Microsoft.Framework.WebEncoders.Tests/HtmlEncoderTests.cs]: I'm admittedly a bit too lazy to look at the product code, but is this unit test just replicating the implementation behavior of the runtime?

File: unicode/Generators/UnicodeTablesGenerator/UnicodeTablesGenerator.csproj [hunk 6]
```diff
@@ -0,0 +1,61 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (Eilon) [inline unicode/Generators/UnicodeTablesGenerator/UnicodeTablesGenerator.csproj]: Any reason not to use project.json for this app?

**Comment by Eilon** [comment]
General question: A lot of the [very extensive!] tests here seem to iterate over jillions of characters each. How long does it take to run all the encoder tests? Seconds? Or minutes? Or more?

File: unicode/Generators/UnicodeTablesGenerator/UnicodeTablesGenerator.csproj (hunk 6 above)
Comment (davidfowl) [inline unicode/Generators/UnicodeTablesGenerator/UnicodeTablesGenerator.csproj]: I was thinking the same thing.

**Comment by GrabYourPitchforks** [comment]
Mere seconds. Most of the loops are only 64k iterations, which the tester can blow through in no time at all.

**Comment by Eilon** [comment]
Ah OK that's cool then. Just wanted to make sure this wouldn't slow the world down.

**Pull request merged**